piptools/warnings.py

```diff
@@ -0,0 +1,29 @@
+from __future__ import absolute_import
+
+import sys
+import warnings
+
+
+class PipToolsBaseWarning:
+    pass
+
+
+class PipToolsPendingDeprecationWarning(PipToolsBaseWarning, PendingDeprecationWarning):
+    """
+    Deprecates things in future versions.
+    """
+
+    pass
+
+
+class PipToolsDeprecationWarning(PipToolsBaseWarning, DeprecationWarning):
+    """
+    Deprecates things in the next or a specific version.
+    """
+
+    pass
+
+
+# Setup if warning filters are not configured
+if not sys.warnoptions:
+    warnings.simplefilter("once", category=PipToolsBaseWarning)
```

Comment:
I'm not a fan of import-time logic execution. You should instead add a function like `setup_warnings()` and call it first thing in `main()`. This way you're more in control of how and when the code is invoked.

Reply:
TBH I had the same thoughts and that was the first implementation with the same naming :D But `pip-tools` doesn't have `main()` or such. In this case, `setup_warnings()` should be called in `piptools.scripts.compile.cli()` and `piptools.scripts.sync.cli()` separately. A few more options in my mind:
- decorator like `@setup_warnings` for `cli()` functions (which is not bad idea)
- refactor cli scripts to have single entry point `main()` and call there `setup_warnings()` (would touch many lines I guess)
- anything more?

With the current implementation main idea is if one day there will be nothing to deprecated for a while, thus there will not be any imports of `piptools.warnings` module (except tests), so warnings filters won't be touched. What do you think?

Reply:
A decorator sounds good. I waw thinking about it too.

Reply:
Okay, let's go ahead with the decorator then.

Reply:
Addressed in 5f8aa2f.